### ISSUE TITLE
No way to append/union a Binary type

### ISSUE THREAD

**Issue by niieani** [rethinkdb/rethinkdb#5209]
There is currently no way to append or do a union with a Binary type. There are many use cases, the one I had in mind is incremental upload / resumable upload. Benefits are many and include monitoring the progress of an upload by pushing out smaller chunks and and appending them to a preexistent document.

I'd imagine this could work:
```js
r.db('test').table('test').get("1").update({ contents: r.row('contents').append(r.binary(...))})
```

// ref @mglukhovsky asked to create this issue during a conversation on Slack //

**Comment by danielmewes** [comment]
We should probably make `add` work for binary.

Marking as a ReQL proposal.

**Comment by niieani** [comment]
Just to add my 2 cents: `add` to me seems a little ambiguous with a binary type, since you can sum binary values when treating them as numbers, hence my proposition of `append` which seems to be the clearest representation of the action.

**Comment by mlucy** [comment]
+1 for using `add` for this, `append` is used to put a single value on the end of an array (`r([1, 2, 3]).append(4)`), not to concatenate two arrays, whereas `add` is used both to concatenate arrays and to concatenate strings (`r([1, 2]).add([3, 4])` and `r("12").add("34")`), which I think is the closest analogy to concatenating binary blobs.

**Comment by niieani** [comment]
Well for the sake of consistency I have to agree with @mlucy, although in my opinion `add` should have not been used for appending to arrays in the first place (concatenating strings seems more intuitive though). After all `number.add(number)` does not concatenate numbers, but `add` them, so the current usage is less intuitive, since there's more to remember about the overloads of the method.

**Comment by mlucy** [comment]
@nileani -- we settled on that behavior because both Python and Ruby use `+` to concatenate arrays.

**Comment by ronzeidman** [comment]
+1
I agree with the proposal, the first thing I tried was "append" when that didn't work I tried "union" only than I remembered that "add" worked for arrays. then I searched for it online.
I think "union" is the best fit.
in the meantime is there a better way than 
`<some binary>.coerceTo('string').add(<someOtherBinary>.coerceTo('string')).coerceTo('binary')`
Are there any bad binaries that cannot be coerced to strings and back?

**Comment by danielmewes** [comment]
@ronzeidman 
> Are there any bad binaries that cannot be coerced to strings and back?

Yes, strings must be valid UTF-8 strings. So some binary data isn't a valid string.

I don't think there currently is any way in ReQL to concatenate binary data.

The original idea behind the binary data type was really just to store small blobs, such as avatar images, but not to do any processing (or incremental updating) on them.
